File: funcscript/core/ExpressionFunction.cs
using System.Diagnostics.SymbolStore;
using System.Runtime.Serialization;
using FuncScript.Core;
using System.Text;
using FuncScript.Error;
using FuncScript.Model;

namespace FuncScript.Core
{
    public class ExpressionFunction : IFsFunction
    {
        private class ParameterDataProvider : IFsDataProvider
        {
            public IParameterList pars;
            public IFsDataProvider parentSymbolProvider;
            public ExpressionFunction expressionFunction;
            public IFsDataProvider ParentProvider => parentSymbolProvider;
            public bool IsDefined(string key)
            {
                return expressionFunction.ParamterNameIndex.ContainsKey(key)
                       || parentSymbolProvider.IsDefined(key);
            }

            public object Get(string name)
            {
                if (expressionFunction.ParamterNameIndex.TryGetValue(name, out var index))
                    return pars.GetParameter(parentSymbolProvider, index);
                return parentSymbolProvider.Get(name);
            }
        }

        public ExpressionBlock Expression { get; set; }

        private Dictionary<string, int> ParamterNameIndex;
        private String[] _parameters;
        private object _expressionValue = null;
        private IFsDataProvider _context = null;

        public void SetContext(IFsDataProvider context)
        {
            //REVIEW: this is commented because it was being triggered but why was it triggered?
            //if (_context != null)
             //   throw new EvaluationTimeException("Context for expression function already set");
            _context = context;
        }
        public ExpressionFunction(String[] pars, ExpressionBlock exp)
        {
            this.Expression = exp;
            this._parameters = pars;
            this.ParamterNameIndex = new Dictionary<String, int>();
            var i = 0;
            foreach (var n in pars)
                this.ParamterNameIndex.Add(n.ToLower(), i++);
        }

        public int MaxParsCount => _parameters.Length;
        public CallType CallType => CallType.Infix;

        public string Symbol => null;

        public int Precedence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (_context == null)
                throw new Error.EvaluationTimeException("Context not set to expression function");
            var ret= Expression.Evaluate(new ParameterDataProvider
            {
                expressionFunction = this,
                parentSymbolProvider = new KvcProvider(_context,parent),
                pars = pars
            });
            return ret;
        }

        

        public string ParName(int index)
        {
            return _parameters[index];
        }

        public override String ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append(this.Symbol);
            sb.Append('(');
            int c = this.MaxParsCount;
            for (int i = 0; i < c; i++)
            {
                if (i > 0)
                    sb.Append(',');
                sb.Append(this.ParName(i));
            }

            sb.Append(')');
            sb.Append("=>");
            sb.Append(this.Expression.AsExpString(new DefaultFsDataProvider()));
            return sb.ToString();
        }
    }

    
}

File: js-port/fs-js/src/core/expression-function.js
const { BaseFunction, CallType } = require('./function-base');
const { KvcProvider, FsDataProvider } = require('./data-provider');
const { ensureTyped } = require('./value');

class ParameterProvider extends FsDataProvider {
  constructor(expressionFunction, parentProvider, parameterList) {
    super(parentProvider);
    this.expressionFunction = expressionFunction;
    this.parameterList = parameterList;
  }

  get(name) {
    const lower = name.toLowerCase();
    const index = this.expressionFunction.parameterIndex.get(lower);
    if (index !== undefined) {
      return ensureTyped(this.parameterList.getParameter(this.parent, index));
    }
    return super.get(name);
  }

  isDefined(name) {
    const lower = name.toLowerCase();
    if (this.expressionFunction.parameterIndex.has(lower)) {
      return true;
    }
    return super.isDefined(name);
  }
}

class ExpressionFunction extends BaseFunction {
  constructor(parameters, expressionBlock) {
    super();
    this.parameters = parameters || [];
    this.expression = expressionBlock;
    this.parameterIndex = new Map();
    this.context = null;

    this.parameters.forEach((name, idx) => {
      this.parameterIndex.set(name.toLowerCase(), idx);
    });
  }

  setContext(context) {
    this.context = context;
  }

  get maxParameters() {
    return this.parameters.length;
  }

  get callType() {
    return CallType.Infix;
  }

  evaluate(provider, parameterList) {
    if (!this.context) {
      throw new Error('Context not set on expression function');
    }
    const parentChain = new KvcProvider(this.context, provider);
    const parameterProvider = new ParameterProvider(this, parentChain, parameterList);
    const result = this.expression.evaluate(parameterProvider);
    return ensureTyped(result);
  }

  parName(index) {
    return this.parameters[index];
  }
}

module.exports = {
  ExpressionFunction
};
